staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_source.go

```diff
@@ -0,0 +1,467 @@
+
+	Evaluator          Evaluator
+	ConfigurationError error
+}
+
+var _ Source[PolicyHook[runtime.Object, runtime.Object]] = &policySource[runtime.Object, runtime.Object]{}
+
+func NewPolicySource[P runtime.Object, B runtime.Object](
+	policyInformer cache.SharedIndexInformer,
+	bindingInformer cache.SharedIndexInformer,
+	newPolicyAccessor func(P) PolicyAccessor,
+	newBindingAccessor func(B) BindingAccessor,
+	compiler func(P) Evaluator,
+	paramInformerFactory informers.SharedInformerFactory,
+	dynamicClient dynamic.Interface,
+	restMapper meta.RESTMapper,
+) Source[PolicyHook[P, B]] {
+	res := &policySource[P, B]{
+		compiler:             compiler,
+		policyInformer:       generic.NewInformer[P](policyInformer),
+		bindingInformer:      generic.NewInformer[B](bindingInformer),
+		compiledPolicies:     map[types.NamespacedName]compiledPolicyEntry{},
+		newPolicyAccessor:    newPolicyAccessor,
+		newBindingAccessor:   newBindingAccessor,
+		paramsCRDControllers: map[schema.GroupVersionKind]*paramInfo{},
+		informerFactory:      paramInformerFactory,
+		dynamicClient:        dynamicClient,
+		restMapper:           restMapper,
+	}
+	return res
+}
+
+func (s *policySource[P, B]) Run(ctx context.Context) error {
+	if s.ctx != nil {
+		return fmt.Errorf("policy source already running")
+	}
+
+	// Wait for initial cache sync of policies and informers before reconciling
+	// any
+	if !cache.WaitForNamedCacheSync(fmt.Sprintf("%T", s), ctx.Done(), s.policyInformer.HasSynced, s.bindingInformer.HasSynced) {
+		err := ctx.Err()
+		if err == nil {
+			err = fmt.Errorf("initial cache sync for %T failed", s)
+		}
+		return err
+	}
+
+	s.ctx = ctx
+
+	// Perform initial policy compilation after initial list has finished
+	s.notify()
+	s.refreshPolicies()
+
+	notifyFuncs := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(_ interface{}) {
+			s.notify()
+		},
+		UpdateFunc: func(_, _ interface{}) {
+			s.notify()
+		},
+		DeleteFunc: func(_ interface{}) {
+			s.notify()
+		},
+	}
+	handle, err := s.policyInformer.AddEventHandler(notifyFuncs)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := s.policyInformer.RemoveEventHandler(handle); err != nil {
+			utilruntime.HandleError(fmt.Errorf("failed to remove policy event handler: %v", err))
+		}
+	}()
+
+	bindingHandle, err := s.bindingInformer.AddEventHandler(notifyFuncs)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := s.bindingInformer.RemoveEventHandler(bindingHandle); err != nil {
+			utilruntime.HandleError(fmt.Errorf("failed to remove binding event handler: %v", err))
+		}
+	}()
+
+	//!TODO: also watch CRDs? Our native type restmapper should be sufficient
+	// for native types. We could avoid restmapper/update issues by looking up
+	// GVK -> GVR via CRDs directly. Only problem is that is kinda complicated
+	// to maintain a CRD restmapper. Since we'd need to keep it updated
+	// (cant just notify and lookup CRD by its GVK. Why are params specified via GVK lol?)
```

Comment:
Removed this comment since current implementation matches pre-existing behavior (have to wait up to 30s for restmapper resync for CRD changes to become visible). 

The change to use a CRD informer is a bit more involved since the Plugin does not have access to the apiextensions shared informer factory. I think we can defer this improvement to another PR?